### ISSUE TITLE
SCSS hot reload is not working

### ISSUE THREAD

**Issue by AdamWillden** [aurelia/hot-module-reload#16]
@EisenbergEffect thanks for doing the release, unfortunately the new feat in #12 doesn't seem to work for me. @niieani, @jods4 did you try it out at the time?

**I'm submitting a bug report**

* **Library Version:**
0.2.0

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
8.2.1

* **NPM Version:**
5.4.2

* **JSPM OR Webpack AND Version**
Webpack 3.6.0

* **Browser:**
Chrome 61.0.3163.100

* **Language:**
TypeScript 2.5.2

**Current behavior:**
All my dependencies are up to date

```json
{
  "version": "1.0.0",
  "sasslintConfig": "node_modules/siopa-package-tools/.sass-lint.yml",
  "private": true,
  "scripts": {
    "test": "jest",
    "start": "webpack"
  },
  "dependencies": {
    "aurelia-bootstrapper": "^2.1.1",
    "aurelia-dependency-injection": "^1.3.2",
    "aurelia-framework": "^1.1.5",
    "aurelia-pal": "^1.4.0",
    "aurelia-router": "^1.4.0"
  },
  "devDependencies": {
    "@types/extract-text-webpack-plugin": "^3.0.0",
    "@types/jest": "^21.1.2",
    "@types/node": "^8.0.32",
    "aspnet-webpack": "^2.0.1",
    "aurelia-hot-module-reload": "^0.2.0",
    "aurelia-polyfills": "^1.2.2",
    "aurelia-webpack-plugin": "^2.0.0-rc.5",
    "css-loader": "^0.28.7",
    "extract-text-webpack-plugin": "^3.0.0",
    "file-loader": "^1.1.5",
    "gulp": "github:gulpjs/gulp#4.0",
    "import-glob-loader": "^1.1.0",
    "node-sass": "^4.5.3",
    "sass-loader": "^6.0.6",
    "siopa-package-tools": "^1.0.9",
    "style-loader": "^0.19.0",
    "ts-loader": "^2.3.7",
    "ts-node": "^3.3.0",
    "tslib": "^1.7.1",
    "tslint": "^5.5.0",
    "typescript": "^2.5.2",
    "webpack": "^3.6.0",
    "webpack-hot-middleware": "^2.19.1"
  }
}
```

Given the following component, when changing the color, an error is displayed in console.
```js
// relevant webpack config line
{ test: /\.scss$/i, issuer: /\.html$/i, loader: 'css-loader!sass-loader' },
```

```html
<template class="c-header-bar">
  <require from="./header-bar.scss"></require>
  My Header Text
</template>
```

```scss
.c-header-bar {
  color: blue;
}
```

```
process-update.js:115 [HMR] Cannot check for update (Full reload needed)
handleError @ process-update.js:115
process-update.js:116 [HMR] TypeError: Cannot read property 'resources' of undefined
    at HmrContext.29.HmrContext.reloadCss (http://localhost:5000/dist/app.js?v=QNzl5O3zgddSQAkYZUauFSbxSVlDVeKMHEZPOYHMqfk:10355:29)
...ETC
handleError @ process-update.js:116
aurelia-hot-module-reload.js:235 Uncaught (in promise) TypeError: Cannot read property 'resources' of undefined
    at HmrContext.29.HmrContext.reloadCss (aurelia-hot-module-reload.js:235)
    at Object.hot (aurelia-hot-module-reload.js:79)
    at aurelia-loader-webpack.js:167
    at Object.hotApply [as apply] (bootstrap 8fa308dbf2e403c4d37c:580)
    at cb (process-update.js:52)
    at process-update.js:68
    at <anonymous>
```

**Expected/desired behavior:**

Given the following component, when changing the color, then the rendered text should change color without console errors.

If I use `.css` files instead with the same setup and following webpack config line everything works

```js
        { test: /\.css$/i, issuer: /\.html$/i, loader: 'css-loader' },
```

**Comment by jods4** [comment]
I did the change in #12 mostly because it's the same code as another PR I did in aurelia-templating-resources, to keep things consistent.
I _think_ I had a basic LESS example working, but to be honest that was long ago and I'm not 100% sure. There has been major Webpack releases since then so even if it worked it might have been broken in the meantime.

I'm confident this _could_ work with HMR but someone needs to debug the error and fix it.

**Comment by AdamWillden** [comment]
Found some time to look into this. The culprit is obvious. It's fixed as a `css-resource-plugin` but the resource is under `scss-resource-plugin`.

![image](https://user-images.githubusercontent.com/4629144/31779063-21599530-b4ea-11e7-855b-7c946860b3e0.png)

Any thoughts on the best way to determine the plugin to look for? I assume the same code exists elsewhere.

**Comment by jods4** [comment]
To clarify:
1. do you transpile your scss into css with a Webpack loader? So at runtime, `css-resource-plugin` _is_ right.
2. or do you pass scss code to the runtime, hoping it will use a converter?

I'm thinking 1. is the right way to do it with Webpack and should work?

**Comment by AdamWillden** [comment]
See the details in the issue above. I think I'm doing 2?? 

----

**FWIW** I just changed the distributed code as follows and it works (just as a test).

From:
```js
        var cssPluginModuleId = this.loader.applyPluginToUrl(moduleId, 'css-resource-plugin');
        console.log("Handling HMR for " + moduleId);
        delete this.loader.moduleRegistry[moduleId];
        delete this.loader.moduleRegistry[cssPluginModuleId];
        var analyzedModule = this.moduleAnalyzerCache["css-resource-plugin!" + moduleId];
```

To:
```js
        var cssPluginModuleId = this.loader.applyPluginToUrl(moduleId, 'scss-resource-plugin');
        console.log("Handling HMR for " + moduleId);
        delete this.loader.moduleRegistry[moduleId];
        delete this.loader.moduleRegistry[cssPluginModuleId];
        var analyzedModule = this.moduleAnalyzerCache[cssPluginModuleId];
```

**Comment by AdamWillden** [comment]
Picture vs thousand words:
![image](https://user-images.githubusercontent.com/4629144/31779986-66df9c2e-b4ec-11e7-8d86-8daf8bab8ca9.png)

**Comment by AdamWillden** [comment]
[The plugin name is convention-based upon the file extension name](https://github.com/aurelia/templating/blob/b43f292afc5ec392461e7bd1774998e7ffd4bfee/src/view-engine.js#L94) so perhaps we can ignorantly build the plugin name within `hot-module-reload` 

...but part of me says there should be an interface we can use in the place where that convention has been implemented. i.e. a `viewEngine.getResourcePluginName` should probably exist

**Comment by jods4** [comment]
Having a static build step with webpack, I don't quite see why you would compile your resources at runtime. Is there a benefit of doing so?

I'm using LESS and I compile everything down to minified CSS before pushing to browser. In my head this is the main/only(?) scenario.

We could try to support both use cases, but that would require more thinking I guess.

**Comment by AdamWillden** [comment]
I guess I misunderstood (and still do misunderstand) what #12 was for if not this. I'm new to webpack and this is part of my migration from JSPM..

I understood the resulting bundle to contain `css` not `scss`. Thus not compiling `scss` at runtime as you seem to suggest. Is the reason for referencing the `scss` not just to trigger the compilation into `css` on change and serve that?  Thats what `'css-loader!sass-loader'` is doing right? 

This is taken from my resulting `app.js` bundle. Which shows `css` and not the variable/mixin containing `scss` that I began with (albeit with `\n` new lines):
```js
// module
exports.push([module.i, ".c-header-bar {\n  align-items: center;\n  display: flex;\n  flex-wrap: wrap;\n  color: #FF5113;\n  -webkit-border-radius: 10px;\n  -moz-border-radius: 10px;\n  border-radius: 10px; }\n  .c-header-bar__logo {\n    flex: 0 0 auto;\n    margin: 1rem; }\n  .c-header-bar__heading {\n    flex: 0 1 auto;\n    margin: 0 1rem; }\n  .c-header-bar__loader {\n    flex: 1 1 auto;\n    font-size: 2em;\n    margin: 0 1rem; }\n  .c-header-bar__nav {\n    flex: 1 0 100%;\n    font-size: 1.1rem;\n    text-align: left; }\n", ""]);
```

Minification is a step I'm yet to do but I wanted to get this working before I complicate things!

Again I'm new to webpack but as far as I can tell I have everything working sensibly.

----

Ah, I've just thought to re-read your earlier question:

>do you transpile your scss into css with a Webpack loader? So at runtime, css-resource-plugin is right.

I was wrong with my earlier reply, which may have confused matters (sorry)... yes, I do transpile my `scss` into `css` with a webpack loader. But the code added in #12 adds a plugin named `scss-resource-plugin` for files loaded with an `.scss` extension. So when the file gets referenced and loaded by aurelia it's loaded with that but the HMR assumes it was loaded with `css-resource-plugin`. 

I don't think I'm doing anything wrong/weird here, I just think this was missed. 

Sorry for wall of text.

**Comment by AdamWillden** [comment]
Having re-read everything again I am definitely doing the method you suggest as the right way of doing things.

**Comment by jods4** [comment]
Yeah, I thought about this a bit more as well.
I'm not even sure aurelia-webpack-plugin supports client-side loaders.

For now I think that if you use webpack to build your project, you should transform everything at build-time and not use loaders at runtime.

Unless a compelling use-case comes up, I think it's better to leave it at that and not support loaders in aurelia-webpack-loader.

Going back to the initial issue: we need to modify hot-module-reload to not apply loader, right? Because right now your css goes through an automatic scss loader because of the extension?

**Comment by AdamWillden** [comment]
@jods4 sorry for delayed response. Ignoring the rest (cause it's not important) I'll focus on this:

> Going back to the initial issue: we need to modify hot-module-reload to not apply loader, right? Because right now your css goes through an automatic scss loader because of the extension?

I'm not sure what you mean by "to not apply loader" or what you're asking in your following question. 

The problem is really simple. As far as I can tell I'm just trying to make it work as you initially intended, you just missed an additional required change. I urge you to look at the proposed fix in #17. The facts are this:

 - The code you added in #12 and https://github.com/aurelia/templating-resources/pull/279 instructs the application to use the existing CSS resource plugin implementation for `less`, `sass`, `scss`, `styl` files
   - As you know, there is no special handling within aurelia for those extensions beyond plain CSS. This is fine because webpack has already transformed the input files to CSS. You said this yourself in [aurelia/templating-resources#279](https://github.com/aurelia/templating-resources/pull/279#issue-215143962)
 - `viewEngine.addResourcePlugin` uses the extension to create a specific resource plugin name
   - Don't be fooled by seeing `scss-resource-plugin`. Any `less/sass/scss/style-resource-plugin` uses the same implementation as the `css-resource-plugin`, just a different instance with a different name. This is the change you created in #12 
 - The code in this repo assumed the plugin name to be `css-resource-plugin` when applying the hot reload. As per the above bullet, this is not true.
   - My fix in #17 removes this assumption and fixes the issue.
